packages/dd-trace/src/llmobs/plugins/vercel-ai.js

```diff
@@ -0,0 +1,475 @@
+  const provider = providerParts?.[0]
+
+  // TODO(sabrenner): explain or simplify this logic
+  switch (provider) {
+    case 'google':
+      return MODEL_PROVIDER_MAPPING[providerParts?.[1]] ?? provider
+    default:
+      return MODEL_PROVIDER_MAPPING[provider] ?? provider
+  }
+}
+
+/**
+ * Safely JSON parses a string value with a default fallback
+ * @param {string} str
+ * @param {any} defaultValue
+ * @returns {Record<string, any> | string | Array<any>}
+ */
+function getJsonStringValue (str, defaultValue) {
+  let maybeValue = defaultValue
+  try {
+    maybeValue = JSON.parse(str)
+  } catch {
+    // do nothing
+  }
+
+  return maybeValue
+}
+
+/**
+ * Get the model metadata from the span tags (top_p, top_k, temperature, etc.)
+ * @param {import('../../../../dd-trace/src/opentracing/span')} span
+ * @returns {Record<string, string>}
+ */
+function getModelMetadata (span) {
+  const modelMetadata = {}
+  for (const metadata of MODEL_METADATA_KEYS) {
+    const metadataTagKey = `gen_ai.request.${metadata}`
+    const metadataValue = getSpanTag(span, metadataTagKey)
+    if (metadataValue) {
+      modelMetadata[metadata] = metadataValue
+    }
+  }
+
+  return modelMetadata
+}
+
+/**
+ * Get the generation metadata from the span tags (maxSteps, maxRetries, etc.)
+ * @param {import('../../../../dd-trace/src/opentracing/span')} span
+ * @returns {Record<string, string>}
+ */
+function getGenerationMetadata (span) {
+  const metadata = {}
+  for (const tag of Object.keys(span.context()._tags)) {
+    if (!tag.startsWith('ai.settings')) continue
+
+    const settingKey = tag.split('.').pop()
+    const transformedKey = settingKey.replaceAll(/[A-Z]/g, letter => '_' + letter.toLowerCase())
+    if (MODEL_METADATA_KEYS.has(transformedKey)) continue
+
+    const settingValue = getSpanTag(span, tag)
+    metadata[settingKey] = settingValue
+  }
+
+  return metadata
+}
+
+class VercelAILLMObsPlugin extends BaseLLMObsPlugin {
+  static get id () { return 'vercel-ai' }
+  static get integration () { return 'vercel-ai' }
+
+  /**
+   * The available tools within the runtime scope of this integration.
+   * This essentially acts as a global registry for all tools made through the Vercel AI SDK.
+   * @type {Set<Record<string, any>>}
+   */
+  #availableTools
+
+  /**
+   * A mapping of tool call IDs to tool names.
+   * This is used to map the tool call ID to the tool name for the output message.
+   * @type {Record<string, string>}
+   */
+  #toolCallIdsToName
+
+  constructor (...args) {
+    super(...args)
+
+    otelSpanStartCh.subscribe(({ ddSpan }) => {
+      if (!isVercelAISpan(ddSpan)) return
+
+      // holding context for llmobs parentage
+      const ctx = {
+        currentStore: { span: ddSpan }
+      }
+
+      const operation = getOperation(ddSpan)
+      const useLlmObsParent = SPANS_TO_USE_LLMOBS_PARENT.has(operation)
+
+      this.start(ctx, {
+        useLlmObsParent,
+        enterIntoLlmObsStorage: false
```

Comment:
Could these be replaced by a different class hierarchy or by runStores/bindStore?

Reply:
this was a bit tricky to me - because we're using the points of the otel spans, which are the start and end, it doesn't leave a lot of flexibility for running/binding stores over a context if the only points of the context we have are disjoint start and end events. but, i could be thinking about this wrong - lmk if so!

since the otel spans started by the vercel ai sdk are activated spans in the proper scope, i wanted to just add an option to use the APM span parentage as opposed to the LLMObs parentage, since these operations should be isolated from whatever manual LLMObs instrumentation users might do.